### PR TITLE
help the ranges concepts recognize standard contiguous iterators in c++14/17

### PR DESCRIPTION
**Pull request by ericniebler**
## Description

In range-v3, @CaseyCarter and I found a sneaky way to "detect" the contiguous iterators of the 3 major stdlib implementations: libstdc++, libc++, and msvc's stdlib. This PR ports the implementation over so that the following standard containers are now recognized as contiguous:

* `std::vector`
* `std::array`
* `std::string`
* `std::string_view`
* `std::span`

this PR also removes a hack in cudax that was only necessary because the std:: containers weren't considered contiguous in c++17.

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h [hunk 1]
```diff
@@ -165,38 +170,74 @@ struct __iter_traits_cache
...
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_const_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_view_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__iterator/iterator_traits.h:211]: I believe this is missing `_Span_iterator`

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h [hunk 2]
```diff
@@ -165,38 +170,74 @@ struct __iter_traits_cache
...
+_LIBCUDACXX_HIDE_FROM_ABI auto
+  __iter_concept_fn(::__gnu_debug::_Safe_iterator<_Ty*, _Range>, __priority_tag<3>) -> contiguous_iterator_tag;
+#endif
+#if defined(__GLIBCXX__)
+_CCCL_TEMPLATE(class _Iter, class _Ty, class _Range)
+_CCCL_REQUIRES(_IsSame<_Iter, ::__gnu_cxx::__normal_iterator<_Ty*, _Range>>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto
+  __iter_concept_fn(::__gnu_cxx::__normal_iterator<_Ty*, _Range>, __priority_tag<3>) -> contiguous_iterator_tag;
+#endif
+#if defined(_LIBCPP_VERSION)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__iterator/iterator_traits.h:185]: I believe those should be `#elif` because we wont mix different STL

**Review by miscco — COMMENTED** [review]
I really did not wanted to do that, but its most likely the easiest solution.

Can you please then fix our `span` to always use the proper range constructor and not the pre-ranges hack?

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h [hunk 3]
```diff
@@ -165,38 +170,77 @@ struct __iter_traits_cache
...
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_const_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_view_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Span_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
```
Comment (CaseyCarter) [inline libcudacxx/include/cuda/std/__iterator/iterator_traits.h:213]: You can probably do this more cleanly in modernish MSVC by detecting if our iterator unwrapping machinery produces a pointer from a given iterator type. `std::is_pointer_v<std::_Unwrapped_t<_Iter>>` is `true` for the STL's contiguous iterators, and for `gsl:span`'s iterators as well. (I didn't tell you to take this dependency on STL internals, I was never here. 😜)

Comment (ericniebler) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:213]: hrm. we need to support older versions of msvc. when did this change go in?

**Comment by ericniebler** [comment]
> Can you please then fix our span to always use the proper range constructor and not the pre-ranges hack?

i started down this rabbit hole, but the range stuff is only defined in C++17, and (for now) `span` still supports C++14.

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h (hunk 3 above)
Comment (CaseyCarter) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:213]: I don't recall precisely when we defined the alias `_Unwrapped_t`, but `is_pointer_v<decltype(::std::_Get_unwrapped(::std::declval<_Iter>()))>` isn't much more to type, it should work back to at least VS2015. CE only goes back to 16.0 (https://www.godbolt.org/z/h4Wa1YvEc), and that works. Is that old enough?

Comment (CaseyCarter) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:213]: Works fine with VS2017 v15.9, not at all with VS2015.

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h [hunk 4]
```diff
@@ -170,24 +170,23 @@ struct __iter_traits_cache
...
-#if defined(__GLIBCXX__)
+#  endif
+#  if defined(__GLIBCXX__)
 _CCCL_TEMPLATE(class _Iter, class _Ty, class _Range)
 _CCCL_REQUIRES(_IsSame<_Iter, ::__gnu_cxx::__normal_iterator<_Ty*, _Range>>::value)
 _LIBCUDACXX_HIDE_FROM_ABI auto
   __iter_concept_fn(::__gnu_cxx::__normal_iterator<_Ty*, _Range>, __priority_tag<3>) -> contiguous_iterator_tag;
-#endif
-#if defined(_LIBCPP_VERSION)
+#  elif defined(_LIBCPP_VERSION)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__iterator/iterator_traits.h:185]: I am sorry, but my comment was erroneous.

We always pull in libstdc++ through some nvcc machinery, so we need to make this a separate branch

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h (hunk 3 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:213]: We only support 2017 gladly

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h [hunk 5]
```diff
@@ -182,36 +183,16 @@ _CCCL_TEMPLATE(class _Iter, class _Ty, class _Range)
...
-#  elif defined(_LIBCPP_VERSION)
+#  endif // __GLIBCXX__
+#  if defined(_LIBCPP_VERSION)
 _CCCL_TEMPLATE(class _Iter, class _Ty)
 _CCCL_REQUIRES(_IsSame<_Iter, ::std::__wrap_iter<_Ty*>>::value)
 _LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(::std::__wrap_iter<_Ty*>, __priority_tag<3>) -> contiguous_iterator_tag;
 #  elif defined(_MSVC_STL_VERSION) || defined(_IS_WRS)
 _CCCL_TEMPLATE(class _Iter)
-_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Array_iterator>::value)
+_CCCL_REQUIRES((_CCCL_TRAIT(is_pointer, ::std::_Unwrapped_t<_Iter>)))
```
Comment (CaseyCarter) [inline libcudacxx/include/cuda/std/__iterator/iterator_traits.h:193]: I was sneaky and changed this to:
```suggestion
_CCCL_REQUIRES((_CCCL_TRAIT(is_pointer, decltype(::std::_Get_unwrapped(::std::declval<_Iter>())))))
```
before validating with older MSVC; VS2017 v15.9 doesn't have the `_Unwrapped_t` alias so we need to expand it.

Do we really need the extra set of `()` around `_CCCL_TRAIT`?

Comment (ericniebler) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:193]: we don't test with VS2017, so i don't think it matters. and no we don't need the extra parens.

File: libcudacxx/include/cuda/std/__iterator/iterator_traits.h (hunk 3 above)
Comment (ericniebler) [reply libcudacxx/include/cuda/std/__iterator/iterator_traits.h:215]: I had to revert the change to use `std::_Unwrapped_t`. something about that formulation was causing msvc to instantiate a template it shouldn't while doing overload resolution.

**Review by miscco — APPROVED** [review]

**Pull request merged**